popde/adaptive_kde.py

```diff
@@ -117,6 +117,100 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+    """
+    Optimize bandwidth and alpha by grid search using
+    cross validation with a log likelihood figure of merit 
+    """
+    def __init__(self, data, bandwidth_options, alpha_options, weights=None, input_transf=None, stdize=False, rescale=None, backend='KDEpy', bandwidth=1.0, alpha=0.0, dim_names=None, do_fit=True, n_splits=2):
+        self.alpha_options = alpha_options
+        self.bandwidth_options = bandwidth_options
+        self.n_splits = n_splits
+        super().__init__(data, weights, input_transf, stdize,
+                         rescale, backend, bandwidth, alpha, dim_names, do_fit)
+
+    def loo_cv_score(self, bandwidth_val, alpha_val):
+        from sklearn.model_selection import LeaveOneOut
+        loo = LeaveOneOut() #sklearn way
+        fom = 0.0
+        for train_index, test_index in loo.split(self.data):
+            train_data, test_data = self.data[train_index], self.data[test_index]
+            #for weights we need corressponding weights as well
+            if self.weights is not None:
+                local_train_weights = weights[train_index]
+            local_weights = None # FIX ME
+            awkde = AdaptiveBwKDE(train_data, local_weights, bandwidth=bandwidth_val,alpha=alpha_val)
+            fom += np.log(awkde.evaluate(test_data))
+        return np.mean(fom)
+
+    def kfold_cv_score(self, bandwidth_val, alpha_val):
+        """
+        Perform k-fold cross-validation with k =n_splits
+        which by default  is 2
+        """
+        from sklearn.model_selection import KFold
+        kf = KFold(n_splits=self.n_splits, shuffle=True, random_state=42)
+        fom = []
+        for train_index, test_index in kf.split(self.data):
+            train_data, test_data = self.data[train_index], self.data[test_index]
+            #for weights we need corressponding weights as well
+            if self.weights is not None:
+                local_train_weights = weights[train_index]
+            local_weights = None # FIX ME
+            awkde = AdaptiveBwKDE(train_data, local_weights, bandwidth=bandwidth_val,alpha=alpha_val)
+            log_kde_eval = np.log(awkde.evaluate(test_data))
+            fom.append(log_kde_eval.sum())
+        return sum(fom)
+
+    def optimize_parameters(self, method='loo_cv', fom_plot_name=None):
+        best_params = {'bandwidth': None, 'alpha': None}
+
+        FOM= {}
+        for bandwidth in self.bandwidth_options:
+            for alpha in self.alpha_options:
+                if method=='kfold_cv':
+                    score = self.kfold_cv_score(bandwidth, alpha)
+                else:
+                    score = self.loo_cv_score(bandwidth, alpha)
+
+                FOM[(bandwidth, alpha)] = score
+
+        optval = max(FOM.items(), key=operator.itemgetter(1))[0]
+        optbw, optalpha  = optval[0], optval[1]
+        best_score = FOM[(optbw, optalpha)]
+        #set self.bandwidth  and self.alpha  as optimized values
+        self.bandwidth  = optbw
```

Comment:
do we not want to use `set_adaptive_parameter` here (method of AdaptiveKDE that both sets the parameters and refits the KDE automatically ?)

Reply:
correct. done